libsrc/histograms_lut/man3/im_stdif.3
.TH IM_STDIF 3 "10 May 1991"
.SH NAME
im_stdif, im_stdif_raw \- statistical differentiation of an image
.SH SYNOPSIS
.B #include <vips/vips.h>

int im_stdif(in, out, a, m0, b, s0, xw, yw)
.br
IMAGE *in, *out;
.br
double a, m0, b, s0;
.br
int xw, yw;

int im_stdif_raw(in, out, a, m0, b, s0, xw, yw)
.br
IMAGE *in, *out;
.br
double a, m0, b, s0;
.br
int xw, yw;

.SH DESCRIPTION
im_stdif() preforms statistical differencing according to the formula
given in page 45 of the book "An Introduction to Digital Image Processing" by
Wayne Niblack. This transformation emphasises the way in which a pel differs
statistically from its neighbours. It is useful for enhancing low-contrast
images with lots of detail, such as X-ray plates.

At point (i,j) the output is given by the eqn:

  vout(i,j) = a*m0 + (1-a)*meanv + 
          (vin(i,j) - meanv) * (b*s0) / (s0+b*stdv)

Values a, m0, b and s0 are entered, while meanv and stdv are the values
calculated over a moving window of size xw, yw centred on pixel (i,j). m0 is the
new mean, a is the weight given to it. s0 is the new standard deviation, b is
the weight given to it. Try:

  im_stdif $VIPSHOME/pics/huysum.v fred.v 0.5 128 0.5 50 11 11

The opreation works on one-band UCHAR images only, and writes a one-band UCHAR
image as its result. The output image has the same size as the input - a black
border is added to mark uncomputable pixels.

im_stdif_raw() behaves exactly as im_stdif(), but does not add the border of
black pixels.

.SH RETURN VALUE
All functions returns 0 on success and -1 on error.
.SH SEE ALSO
im_lhisteq(3), im_histgr(3), im_hsp(3), im_heq(3).
.SH COPYRIGHT
1991-1996, The National Gallery and Birkbeck College


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/histograms_lut/im_stdif.c
/* @(#) Functions which calculates statistical differenciating according to
 * @(#) the formula given in page 45 of the book "An intro to digital image
 * @(#) processing" by Wayne Niblack
 * @(#) 
 * @(#) At point (i,j) the output is given by the eqn:
 * @(#)
 * @(#) 	vout(i,j) = a*m0 +(1-a)*meanv + 
 * @(#) 		(vin(i,j) - meanv) * beta*sigma0/(sigma0+beta*stdv)
 * @(#)
 * @(#) Values a, m0, beta and sigma0 are entered 
 * @(#) meanv and stdv are the values calculated over a moving window
 * @(#) xwin and ywin are the sizes of the used window
 * @(#) The resultant coefficients are written as floats
 * @(#) in out which has a size of in
 * @(#)
 * @(#) int im_stdif(in, im, alpha, mean0, beta, sigma0, xwin, ywin)
 * @(#) IMAGE *in, *out;
 * @(#) int xwin, ywin;
 * @(#) double alpha, mean0, beta, sigma0;
 * @(#)
 * @(#) Returns 0 on sucess  and -1 on error.
 *
 * Copyright: 1990, N. Dessipris.
 *
 * Author: Nicos Dessipris
 * Written on: 02/05/1990
 * Modified on : 
 * 6/8/93 JC
 *	- now works for odd window sizes
 *	- ANSIfication
 * 25/5/95 JC
 *	- new IM_ARRAY() macro
 * 25/1/96 JC
 *	- im_lhisteq() adapted to make new im_stdif()
 *	- now partial, plus rolling window
 *	- 5x faster, amazingly
 *	- works
 * 7/4/04 
 *	- now uses im_embed() with edge stretching on the input, not
 *	  the output
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Hold global stuff here.
 */
typedef struct {
	int xwin, ywin;		/* Parameters */
	double a, m0, b, s0;
} StdifInfo;

/* stdif generate function.
 */
static int
stdif_gen( REGION *or, REGION *ir, IMAGE *in, StdifInfo *inf )
{
	Rect irect;

	Rect *r = &or->valid;
	int le = r->left;
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int ri = IM_RECT_RIGHT(r);

	int x, y, i, j;
	int lsk;

	int coff;			/* Offset to move to centre of window */
	int npel = inf->xwin * inf->ywin;

	/* What part of ir do we need?
	 */
	irect.left = or->valid.left;
	irect.top = or->valid.top;
	irect.width = or->valid.width + inf->xwin;
	irect.height = or->valid.height + inf->ywin;
	if( im_prepare( ir, &irect ) )
		return( -1 );

	lsk = IM_REGION_LSKIP( ir );
	coff = lsk * (inf->ywin/2) + inf->xwin/2;

	for( y = to; y < bo; y++ ) {
		/* Get input and output pointers for this line.
		 */
		PEL *p = (PEL *) IM_REGION_ADDR( ir, le, y );
		PEL *q = (PEL *) IM_REGION_ADDR( or, le, y );
		PEL *p1, *p2;
		int sum = 0;
		int sum2 = 0;

		/* Precompute some factors.
		 */
		double f1 = inf->a * inf->m0;
		double f2 = 1.0 - inf->a;
		double f3 = inf->b * inf->s0;

		/* Find sum, sum of squares for the start of this line.
		 */
		for( p1 = p, j = 0; j < inf->ywin; j++, p1 += lsk )
			for( p2 = p1, i = 0; i < inf->xwin; i++, p2++ ) {
				int t = *p2;

				sum += t;
				sum2 += t * t;
			}

		/* Loop for output pels.
		 */
		for( x = le; x < ri; x++, p++ ) {
			/* Find stats.
			 */
			double mean = (double)sum / npel;
			double var = (double)sum2 / npel - (mean * mean);
			double sig = sqrt( var );

			/* Transform.
			 */
			double res = f1 + f2*mean + ((double) p[coff] - mean) * 
				(f3 / (inf->s0 + inf->b*sig));
			
			/* And write.
			 */
			if( res < 0.0 )
				*q++ = 0;
			else if( res >= 256.0 )
				*q++ = 255;
			else
				*q++ = res + 0.5;

			/* Adapt sums - remove the pels from the left hand
			 * column, add in pels for a new right-hand column.
			 */
			for( p1 = p, j = 0; j < inf->ywin; j++, p1 += lsk ) {
				int t1 = p1[0];
				int t2 = p1[inf->xwin];

				sum -= t1;
				sum2 -= t1 * t1;

				sum += t2;
				sum2 += t2 * t2;
			}
		}
	}

	return( 0 );
}

int 
im_stdif_raw( IMAGE *in, IMAGE *out, 
	double a, double m0, double b, double s0, 
	int xwin, int ywin )
{
	StdifInfo *inf;

	if( m0 < 0 || m0 > 255 || a < 0 || a > 1.0 || b < 0 || b > 2 || 
		s0 < 0 || s0 > 255 ) {
		im_errormsg( "im_stdif: parameters out of range" );
		return( -1 );
	}
	if( im_piocheck( in, out ) )
		return( -1 );
	if( in->Bbits != IM_BBITS_BYTE || in->BandFmt != IM_BANDFMT_UCHAR || 
		in->Bands != 1 || in->Coding != IM_CODING_NONE ) { 
		im_errormsg( "im_stdif: one band uchar uncoded only" ); 
		return( -1 ); 
	}
	if( xwin > in->Xsize || ywin > in->Ysize ) {
		im_errormsg( "im_stdif: window too large" );
		return( -1 );
	}
	if( im_cp_desc( out, in ) ) 
		return( -1 );
	out->Xsize -= xwin;
	out->Ysize -= ywin;

	/* Save parameters.
	 */
	if( !(inf = IM_NEW( out, StdifInfo )) )
		return( -1 );
	inf->xwin = xwin;
	inf->ywin = ywin;
	inf->a = a;
	inf->m0 = m0;
	inf->b = b;
	inf->s0 = s0;

	/* Set demand hints. FATSTRIP is good for us, as THINSTRIP will cause
	 * too many recalculations on overlaps.
	 */
	if( im_demand_hint( out, IM_FATSTRIP, in, NULL ) )
		return( -1 );

	/* Write the hist.
	 */
	if( im_generate( out,
		im_start_one, stdif_gen, im_stop_one, in, inf ) )
		return( -1 );

	return( 0 );
}

/* The above, with a border to make out the same size as in.
 */
int 
im_stdif( IMAGE *in, IMAGE *out, 
	double a, double m0, double b, double s0, 
	int xwin, int ywin )
{
	IMAGE *t1 = im_open_local( out, "im_stdif:1", "p" );

	if( !t1 || 
		im_embed( in, t1, 1, xwin / 2, ywin / 2, 
			in->Xsize + xwin - 1, 
			in->Ysize + ywin - 1 ) ||
		im_stdif_raw( t1, out, a, m0, b, s0, xwin, ywin ) )
		return( -1 );

	return( 0 );
}
